### PR TITLE
#80 pass btn color & vanish after clicked

### LINKED ISSUE #80: Make it optically more obvious whether the "Pass" button has been clicked
* Change it to a fainter grey?
*  Maybe make the button vanish after it has been clicked?

### PR DESCRIPTION
**Pull request by matthewejones**
Changes the pass button colour to make it more obvious and makes it vanish after clicking.
Also reminds you if you're the last one left to pass.

Resolves #80

### CODE REVIEW

**Comment by ChristophNiehoff** [comment]
@JonathanDeul  Is something missing here? I don't see how the last player gets reminded...

**Comment by ChristophNiehoff** [comment]
Also: Immediately after the game was created and before the first card was played, there is a strange horizontal line/shadow where usually the pass button is...  Could you also remove this?

Screenshot from 2021-07-13 21-45-39

**Comment by JonathanDeul** [comment]
> 
> 
> @JonathanDeul Is something missing here? I don't see how the last player gets reminded...

yeah, changes in sidebar.js were somehow not applied. I will fix this tmrw.

**Comment by ChristophNiehoff** [comment]
@JonathanDeul You don't have to create a new PR. You can just update the existing one, i.e. add commits to it.
Could you cherry-pick commits a6a1259 and ad7c923 onto this branch `passbutton`? Then we can close PR #92 as duplicate.

File: src/client/components/sidebar/sidebar.js [hunk 1]
```diff
@@ -29,8 +30,19 @@ class Sidebar extends React.Component {
 
   render() {
     let dealtCard = getDealtCard(this.props.G);
+    let all = new Set(getPlayers(this.props.ctx.numPlayers));
+    let passed = new Set(this.props.G.passed);
+    let difference = new Set([...all].filter(x => !passed.has(x)));
+    let whopassed = resolvePlayerNames(Array.from(difference), this.props.names, this.props.playerID);
+    let btncolor = "secondary", warningtxt = "invisible";
+
+    if (whopassed[0] === "You" && whopassed[1] === undefined) {
```
Comment (ChristophNiehoff) [inline src/client/components/sidebar/sidebar.js:39]: I'm not sure I understand what is happening here:  If `You` are the only one who passed, then show the warning that you are last one to pass? Maybe the variable names need a bit of refinement?

File: src/client/components/sidebar/sidebar.js [hunk 2]
```diff
@@ -43,12 +55,13 @@ class Sidebar extends React.Component {
           <hr />
           <Leaderboard playerID={this.props.playerID} scores={this.props.G.scores} names={this.props.names} cards={getDealtCardsForPlayers(this.props.G.order, this.props.G.dealt)} />
           <hr />
-          <Button color="secondary" size="lg" block disabled={
+          <p class={warningtxt}>You are the last one to pass!</p>
```
Comment (ChristophNiehoff) [inline src/client/components/sidebar/sidebar.js:58]: I believe that generally it is more readable to write this as 
```js
const isLastToPass = whopassed[0] === "You" && whopassed[1] === undefined;

<p className={(isLastToPass) ? "visible" : "invisible"} > ... </p>
```

Comment (ChristophNiehoff) [reply src/client/components/sidebar/sidebar.js:58]: Also the property is called `className`, not `class`.

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]

**Comment by ChristophNiehoff** [comment]
@JonathanDeul  For me, the horizontal line is still present.

Also we now have two different styles for the buttons. That's the appearance (i.e. the rounded corners) as well as the hover effects. I think there should be one consistent style for all buttons.
Screenshot from 2021-07-15 13-57-19

And this was probably unwanted, but your CSS changes also had an effect on the Proceed button on the create-page:
Screenshot from 2021-07-15 13-59-43

**Comment by JonathanDeul** [comment]
> 
> 
> @JonathanDeul For me, the horizontal line is still present.
> 
> Also we now have two different styles for the buttons. That's the appearance (i.e. the rounded corners) as well as the hover effects. I think there should be one consistent style for all buttons.
> Screenshot from 2021-07-15 13-57-19
> 
> And this was probably unwanted, but your CSS changes also had an effect on the Proceed button on the create-page:
> Screenshot from 2021-07-15 13-59-43

Should I change the pass button back to the original appearance or also change the other buttons?

File: src/client/components/sidebar/sidebar.js [hunk 3]
```diff
@@ -42,13 +49,13 @@ class Sidebar extends React.Component {
...
-          <hr />
-          <Button color="secondary" size="lg" block disabled={
+          <p className={(isLastToPass) ? "visible" : "invisible"} >You are the last one to pass!</p>
+          <Button color={(isLastToPass) ? "warning" : "secondary"} size="lg" block disabled={
               this.props.ctx.phase !== "threats" ||
               this.props.G.passed.includes(this.props.playerID) ||
               !this.props.active
             } onClick={() => { this.props.moves.pass() }}>
-              Pass
+              PASS
```
Comment (ChristophNiehoff) [inline src/client/components/sidebar/sidebar.js:58]: I think we should go back to `Pass` as this is how buttons are written everywhere else

File: src/client/components/sidebar/sidebar.js [hunk 4]
```diff
@@ -42,13 +49,13 @@ class Sidebar extends React.Component {
           </Button>
           <hr />
           <Leaderboard playerID={this.props.playerID} scores={this.props.G.scores} names={this.props.names} cards={getDealtCardsForPlayers(this.props.G.order, this.props.G.dealt)} />
-          <hr />
-          <Button color="secondary" size="lg" block disabled={
+          <p className={(isLastToPass) ? "visible" : "invisible"} >You are the last one to pass!</p>
```
Comment (ChristophNiehoff) [inline src/client/components/sidebar/sidebar.js:52]: As written in the other PR, it's better to use conditional rendering than making it invisible via CSS.

Comment (ChristophNiehoff) [reply src/client/components/sidebar/sidebar.js:52]: Could you also move this line a little bit down? It looks to be a bit to high.
Screenshot from 2021-07-15 22-24-35
And maybe use a color that is better readable on a white background?

File: src/client/components/sidebar/sidebar.js [hunk 5]
```diff
@@ -29,8 +30,14 @@ class Sidebar extends React.Component {
 
   render() {
     let dealtCard = getDealtCard(this.props.G);
+    let all = new Set(getPlayers(this.props.ctx.numPlayers));
+    let passed = new Set(this.props.G.passed);
+    let difference = new Set([...all].filter(x => !passed.has(x)));
+    let whopassed = resolvePlayerNames(Array.from(difference), this.props.names, this.props.playerID);
+    const isLastToPass = whopassed[0] === "You" && whopassed[1] === undefined;
```
Comment (ChristophNiehoff) [inline src/client/components/sidebar/sidebar.js:37]: Would it work to define this as 
```js
const isLastToPass = this.props.G.passed.length === this.props.ctx.numPlayers-1 && !this.props.G.passed.includes(this.props.playerID)
```
?

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]
I really like this. Together with the checkmarks this is really awesome!

I just added a few coding style comments

**Review by ChristophNiehoff — APPROVED** [review]

**Pull request merged**